cmd/cloud-controller-manager/nodeipamcontroller.go

```diff
@@ -180,50 +155,80 @@ func processCIDRs(cidrsList string) ([]*net.IPNet, bool, error) {
 // for --node-cidr-mask-size-ipv4 and --node-cidr-mask-size-ipv6 respectively. If value not provided,
 // then it will return default IPv4 and IPv6 cidr mask sizes.
-func setNodeCIDRMaskSizesDualStack(cfg nodeipamconfig.NodeIPAMControllerConfiguration) (int, int, error) {
-	ipv4Mask, ipv6Mask := defaultNodeMaskCIDRIPv4, defaultNodeMaskCIDRIPv6
-	// NodeCIDRMaskSize can be used only for single stack clusters
-	if cfg.NodeCIDRMaskSize != 0 {
-		return ipv4Mask, ipv6Mask, errors.New("usage of --node-cidr-mask-size is not allowed with dual-stack clusters")
+func setNodeCIDRMaskSizes(cfg nodeipamconfig.NodeIPAMControllerConfiguration, clusterCIDRs []*net.IPNet) ([]int, error) {
+
+	sortedSizes := func(maskSizeIPv4, maskSizeIPv6 int) []int {
+		nodeMaskCIDRs := make([]int, len(clusterCIDRs))
+
+		for idx, clusterCIDR := range clusterCIDRs {
+			if netutils.IsIPv6CIDR(clusterCIDR) {
+				nodeMaskCIDRs[idx] = maskSizeIPv6
+			} else {
+				nodeMaskCIDRs[idx] = maskSizeIPv4
+			}
+		}
+		return nodeMaskCIDRs
 	}
-	if cfg.NodeCIDRMaskSizeIPv4 != 0 {
-		ipv4Mask = int(cfg.NodeCIDRMaskSizeIPv4)
+
+	// --node-cidr-mask-size flag is incompatible with dual stack clusters.
+	ipv4Mask, ipv6Mask := defaultNodeMaskCIDRIPv4, defaultNodeMaskCIDRIPv6
+	isDualstack := len(clusterCIDRs) > 1
+
+	// case one: cluster is dualstack (i.e, more than one cidr)
+	if isDualstack {
+		// if --node-cidr-mask-size then fail, user must configure the correct dual-stack mask sizes (or use default)
+		if cfg.NodeCIDRMaskSize != 0 {
+			return nil, errors.New("usage of --node-cidr-mask-size is not allowed with dual-stack clusters")
+
+		}
+
+		if cfg.NodeCIDRMaskSizeIPv4 != 0 {
+			ipv4Mask = int(cfg.NodeCIDRMaskSizeIPv4)
+		}
+		if cfg.NodeCIDRMaskSizeIPv6 != 0 {
+			ipv6Mask = int(cfg.NodeCIDRMaskSizeIPv6)
+		}
+		return sortedSizes(ipv4Mask, ipv6Mask), nil
 	}
-	if cfg.NodeCIDRMaskSizeIPv6 != 0 {
-		ipv6Mask = int(cfg.NodeCIDRMaskSizeIPv6)
+
+	maskConfigured := cfg.NodeCIDRMaskSize != 0
+	maskV4Configured := cfg.NodeCIDRMaskSizeIPv4 != 0
+	maskV6Configured := cfg.NodeCIDRMaskSizeIPv6 != 0
+	isSingleStackIPv6 := netutils.IsIPv6CIDR(clusterCIDRs[0])
+
+	// original flag is set
+	if maskConfigured {
+		// original mask flag is still the main reference.
+		if maskV4Configured || maskV6Configured {
+			return nil, errors.New("usage of --node-cidr-mask-size-ipv4 and --node-cidr-mask-size-ipv6 is not allowed if --node-cidr-mask-size is set. For dual-stack clusters please unset it and use IPFamily specific flags")
+		}
+
+		mask := int(cfg.NodeCIDRMaskSize)
+		return sortedSizes(mask, mask), nil
 	}
-	return ipv4Mask, ipv6Mask, nil
-}
 
-// getNodeCIDRMaskSizes is a helper function that helps the generate the node cidr mask
-// sizes slice based on the cluster cidr slice
-func getNodeCIDRMaskSizes(clusterCIDRs []*net.IPNet, maskSizeIPv4, maskSizeIPv6 int) []int {
-	nodeMaskCIDRs := make([]int, len(clusterCIDRs))
+	if maskV4Configured {
```

Comment:
This feels a little convoluted to me.  I think one of three changes.

Either a) Don't worry about "extra" flags:

```
	if maskV4Configured {
		ipv4Mask = int(cfg.NodeCIDRMaskSizeIPv4)
	}
	if maskV6Configured {
		ipv6Mask = int(cfg.NodeCIDRMaskSizeIPv6)
	}
	return sortedSizes(ipv4Mask, ipv6Mask), nil
```

Yeah, the user might pass a v6 flag on a v4 cluster, but is that really a
problem?  We won't use it thanks to `sortedSizes`.

Or b) Clarify:

```
	if maskV4Configured && isSingleStackIPv6 {
		return nil, errors.New("usage of --node-cidr-mask-size-ipv4 is not allowed for a single-stack IPv6 cluster")
	} else if maskV6Configured  && !isSingleStackIPv6 {
		return nil, errors.New("usage of --node-cidr-mask-size-ipv6 is not allowed for a single-stack IPv4 cluster")
	}

	if maskV4Configured {
		ipv4Mask = int(cfg.NodeCIDRMaskSizeIPv4)
	}
	if maskV6Configured {
		ipv6Mask = int(cfg.NodeCIDRMaskSizeIPv6)
	}

	return sortedSizes(ipv4Mask, ipv6Mask), nil
```

Or c) Simplify (closer to what's here:

```
	if maskV4Configured {
		if isSingleStackIPv6 {
			return nil, errors.New("usage of --node-cidr-mask-size-ipv4 is not allowed for a single-stack IPv6 cluster")
		}
		ipv4Mask = int(cfg.NodeCIDRMaskSizeIPv4)
	} else if maskV6Configured {
		if !isSingleStackIPv6 {
			return nil, errors.New("usage of --node-cidr-mask-size-ipv6 is not allowed for a single-stack IPv4 cluster")
		}
		ipv6Mask = int(cfg.NodeCIDRMaskSizeIPv6)
	}

	return sortedSizes(ipv4Mask, ipv6Mask), nil
```

In this last form I just removed the 2nd sub-block of the first case.

Reply:
The entire purpose of `too many checks` is to ensure that user does not use a flag that is ignored case A+B above. 

Case C is the same. If the user used two flags (IPv4 and IPv6) the question comes in forms?
-- did you really want to use dual stack but didn't set two cidrs?
-- did you choose the right family for your cluster (once NodeIPAM starts it will set immutable values).

These validation are on flags that does not really change, having errors clear and pedantic helps the user to make the right decision from early specially when the decision will result in hard to fix data.

if you push i will go for case C. but i urge you not to.